internal/exporter/exporter_test.go

```diff
@@ -0,0 +1,163 @@
+package exporter_test
+
+import (
+	"archive/tar"
+	"bytes"
+	"io"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/ethersphere/bee-repair/internal/exporter"
+	"github.com/ethersphere/bee/pkg/shed"
+	chunktesting "github.com/ethersphere/bee/pkg/storage/testing"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+type checkUpdater struct {
+	prev int
+	t    *testing.T
+}
+
+func (c *checkUpdater) Update(done, total int) {
+	if c.prev > done {
+		c.t.Fatal("update arrive with older progress")
+	}
+	if done > total {
+		c.t.Fatal("incorrect update")
+	}
+	c.prev = done
+}
+
+func TestExporter(t *testing.T) {
+	verifyTar := func(t *testing.T, tr *tar.Reader, chunkMap map[string]swarm.Chunk) {
+		for {
+			hdr, err := tr.Next()
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				t.Fatal(err)
+			}
+			if hdr.Name == exporter.ExportVersionFilename {
+				if hdr.Size != int64(len(exporter.CurrentExportVersion)) {
+					t.Fatal("invalid header size for version entry")
+				}
+				continue
+			}
+
+			chunk, found := chunkMap[hdr.Name]
+			if !found {
+				t.Fatalf("chunk %s not found", hdr.Name)
+			}
+
+			if hdr.Size != int64(len(chunk.Data())) {
+				t.Fatalf("invalid chunksize, expected %d got %d", len(chunk.Data()), hdr.Size)
+			}
+
+			limitRd := &io.LimitedReader{
+				R: tr,
+				N: hdr.Size,
+			}
+			chunkBuf := make([]byte, hdr.Size)
+			_, err = limitRd.Read(chunkBuf)
+			if err != nil && err != io.EOF {
+				t.Fatal(err)
+			}
+			if !bytes.Equal(chunk.Data(), chunkBuf) {
+				t.Fatal("invalid data bytes")
+			}
+		}
+	}
+
+	t.Run("default", func(t *testing.T) {
+		defer os.RemoveAll("src")
+		defer os.RemoveAll(filepath.Join(".", exporter.DefaultExportFilename))
+
+		err := os.Mkdir("src", 0775)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		chMap, err := createTestStore("src")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		err = exporter.Export("src")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		tarFile, err := os.Open(exporter.DefaultExportFilename)
+		if err != nil {
+			t.Fatal(err)
+		}
+		tr := tar.NewReader(tarFile)
+
+		verifyTar(t, tr, chMap)
+
+	})
+	t.Run("options", func(t *testing.T) {
+		testFileName := "testexportfile.tar"
+		defer os.RemoveAll("src")
+		defer os.RemoveAll(filepath.Join(".", testFileName))
+
+		err := os.Mkdir("src", 0775)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		chMap, err := createTestStore("src")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		updater := &checkUpdater{t: t}
+		err = exporter.Export(
+			"src",
+			exporter.WithDestinationFilename(testFileName),
+			exporter.WithProgressUpdater(updater),
+		)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if updater.prev != 100 {
+			t.Fatal("Final update incorrect")
+		}
+
+		tarFile, err := os.Open(filepath.Join(".", testFileName))
+		if err != nil {
+			t.Fatal(err)
+		}
+		tr := tar.NewReader(tarFile)
+
+		verifyTar(t, tr, chMap)
+
+	})
+}
+
+func createTestStore(src string) (map[string]swarm.Chunk, error) {
```

Comment:
I would even try to create the test with the actual bee localstore to test the compatibility between GetRetrievalIndex and a specific localstore implementation.

Reply:
Can you elaborate a bit?

Should I move the implementation to this repo? Or should I use an actual DB file?

Reply:
I think that it is ok to import localstore from bee, construct the instance and put chunks into it in a separate test. This can be done in a followup PR, also. That test would ensure that the exporter works with the localstore created indexes. It is more up to you to see if you se fit for such test to be added. This is just my suggestion.

Reply:
@janos the database schema changed and the export version also incremented on `master`, that is why this PR is needed. We could theoretically support that inside the localstore code, but I think that this is something which can be eliminated at this stage of the product lifecycle